### PR TITLE
Refactor datafusion/src/physical_plan/common.rs build_file_list to take less param and reuse code

### LINKED ISSUE #254: Refactor datafusion/src/physical_plan/common.rs build_file_list to take less param and reuse code
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
(This section helps Arrow developers understand the context and *why* for this feature, in addition to  the *what*)

So far the `build_file_list` fn takes a mutable Vec and requires all call sites to pass in the param. It is a better API if the recursive fn call is wrapped.

**Describe the solution you'd like**

Solution is in https://github.com/apache/arrow-datafusion/pull/253

**Describe alternatives you've considered**

Alternative is to leave it as is. But this way we'll have multiple code duplicates.

**Additional context**

### PR DESCRIPTION
**Pull request by Jimexist**
# Which issue does this PR close?

We generally require a GitHub issue to be filed for all bug fixes and enhancements and this helps us generate change logs for our releases. You can link an issue to this PR using the GitHub syntax. For example `Closes #123` indicates that this PR will close issue #123.

Closes #254

 # Rationale for this change
 Why are you proposing this change? If this is already explained clearly in the issue then this section is not needed.
 Explaining clearly why changes are proposed helps reviewers understand your changes and offer better suggestions for fixes.  

# What changes are included in this PR?

There is no need to duplicate the description in the issue here but it is sometimes worth providing a summary of the individual changes in this PR.

Basically a new fn without the mut ref accumulator list param is added, wrapping the original one. All call sites are updated.

# Are there any user-facing changes?

- No

If there are user-facing changes then we may require documentation to be updated before approving the PR.

If there are any breaking changes to public APIs, please add the `breaking change` label.

### CODE REVIEW

File: datafusion/src/physical_plan/hash_join.rs [hunk 1]
```diff
@@ -389,7 +389,7 @@ impl ExecutionPlan for HashJoinExec {
             num_output_rows: 0,
             join_time: 0,
             random_state: self.random_state.clone(),
-            visited_left_side: visited_left_side,
+            visited_left_side,
```
Comment (Jimexist) [inline datafusion/src/physical_plan/hash_join.rs:392]: this fixes a clippy warning

Comment (alamb) [reply datafusion/src/physical_plan/hash_join.rs:392]: I just noticed the same thing  https://github.com/apache/arrow-datafusion/pull/259

**Review by alamb — APPROVED** [review]
Thanks @Jimexist  -- looks like a nice improvement to me.

This is likely technically a "breaking change" (as the function is `pub`) but I think it is minor and not likely to be a huge problem

**Review by Dandandan — APPROVED** [review]
Nice idea @Jimexist
